### PR TITLE
mark all device functions as convergent

### LINKED ISSUE #894: __ockl_lane_u32 should not be moved out of divergent loop
This is a copy of the internal issue SWDEV-167691.

In the attached file, the loop condition depends on __ockl_lane_u32() which itself expands to mbcnt_lo and mbcnt_hi. These intrinsics are marked as sources of divergence, but they get moved out of the loop, thereby incorrectly providing a constant value to the loop condition.

[waterfall_converged.cpp.txt](https://github.com/RadeonOpenCompute/hcc/files/2461675/waterfall_converged.cpp.txt)
[waterfall_converged-gfx900.isa.txt](https://github.com/RadeonOpenCompute/hcc/files/2461676/waterfall_converged-gfx900.isa.txt)

### PR DESCRIPTION
**Pull request by ssahasra**
This fixes https://github.com/RadeonOpenCompute/hcc/issues/894

The intrinsics mbcnt.lo and mbcnt.hi are marked as divergent, but this is not sufficient to prevent them from being inferred as loop-invariant. Marking a function as convergent is the only way to disable LLVM optimizations around it. This is already done by Clang for OpenCL, and HCC needs a place to do the same. The SelectAcceleratorCode pass is in the right place for this.

### CODE REVIEW

**Comment by b-sumner** [comment]
Where in clang are all OpenCL builtin functions marked convergent?  opencl-c.h doesn't make them all convergent.

**Review by scchan — CHANGES_REQUESTED** [review]
It is too big of a hammer to mark all functions as convergent and I don't think this is the right way to fix issues with mbcnt.lo/hi

**Comment by ssahasra** [comment]
This issue is not limited to mbcnt.lo/hi. I can build a different example with __builtin_amdgcn_read_exec(), which produces a similar result. In general, the "convergent" attribute seems to have taken on a wider meaning in LLVM than originally intended. Whereas one would expect it to describe truly control-independent operations like barriers, it is now used to describe all sorts of control-sensitive operations that should not be touched by optimizations. In practice, "convergent" seems to imply any of the following, or more:

1. produces divergence, or
2. must be executed convergently by the whole wave, or
3. returned value depends on execution  mask

Clang conservatively marks **all device functions** as convergent, as seen here:
https://github.com/llvm-mirror/clang/blob/master/lib/CodeGen/CGCall.cpp#L1792

```
  if (getLangOpts().assumeFunctionsAreConvergent()) {
    // Conservatively, mark all functions and calls in CUDA and OpenCL as
    // convergent (meaning, they may call an intrinsically convergent op, such
    // as __syncthreads() / barrier(), and so can't have certain optimizations
    // applied around them).  LLVM will remove this attribute where it safely
    // can.
    FuncAttrs.addAttribute(llvm::Attribute::Convergent);
  }
```
So this is already happening for OpenCL, HIP and CUDA. My patch brings HCC to the fold by performing this in SAC, since that is where we recognize device code.

I agree this is a big hammer, but at least it just re-establishes current practices, which is sufficient for unblocking my actual work. I would like to revisit this whole "convergent" attribute business separately. A cleaner solution will require changes to all LLVM optimizations that currently check for this attribute.

A good example of this behaviour is to look at the compiled device libraries, where all functions are marked as convergent in the .bc file, but not in the source .cl file.

**Comment by AlexVlx** [comment]
I don't know that I feel strongly about this either way, except in that it points out yet another limitation with the "LALA no SIMD to see here, move along! (after some time) right, how do I fix this up to actually work with the hardware that I have:-s" model we've embraced - the GCC work may have a point. It's also unclear if the choice made for CUDA / OCL was deeply thought out or it was a case of a hacky workaround being ritualistically copied once confined to genetic memory. 

That being said, I think that squatting on SAC is problematic on two accounts. First, just plopping lots of things in there because it's convenient goes against the idea of one responsibility per-pass (in hindsight, I probably should not have gone with adding AlwaysInline there). Second, and more importantly, SAC is getting sacked, and not here to stay, therefore all of these added things will have to find a new home. Finally, observe that Clang also concerns itself with marking ASM blocks as convergent, which I believe will be missed if you do it this way.

TL;DR, if we really want to match what Clang does, let's just do that, by changing assumeFunctionsAreConvergent() to return true for HC mode as well (this would be identified by the conjunction `CPlusPlusAMP && DevicePath`. As a bonus that also makes it obvious that you're just adopting common Clang practice.

**Comment by ssahasra** [comment]
Oh, cool! I never thought of looking to see if Clang was aware of the device/host distinction in HCC. That was mainly because of the existence of SAC ... if DevicePath is reliable, then that is the cleanest way to resolve this issue. For now. I agree with the part about SIMT semantics until way later. Like I said, the "convergent" attribute is the real hammer here ... it's being used to protect all of SIMT semantics from optimizations that assume a single-threaded execution.

**Comment by arsenm** [comment]
This shouldn't be done in an IR pass, clang needs to do this correctly upfront

**Comment by ssahasra** [comment]
I just tried this in Clang, and it turns out that it's not enough. Clang correctly marks __ockl_lane_id() as convergent, but then later SAC inlines this function which exposes mbcnt. This is never seen by Clang, and there is no other pass that marks this as convergent. The root cause is that "convergent" is being misused. Ideally, mbcnt should be marked as "control-sensitive" or whatever, and Loop transforms in LLVM should recognize this.

For now, the only workaround that I see is to keep SAC around a bit longer and make it emulate Clang's behaviour. I really need this to unblock my work.

**Comment by arsenm** [comment]
I don’t understand what you mean. hcc should just mark all functions in all circumstances as convergent and ignore worrying if it’s a device function?

**Comment by ssahasra** [comment]
The pass "SelectAcceleratorCode" or "SAC" is designed to keep only device functions and remove everything else. My change teaches that pass to also mark all device functions as convergent. At this point in the IR, all referenced intrinsics are also present as declarations and they will get marked too. This is only a bigger hammer to cover the parts that weren't visible to earlier "convergent" hammer in Clang.

**Comment by ssahasra** [comment]
@AlexVlx, you mentioned that SAC is going away. Then at what point in the IR can I assume that non-device functions have been filtered out? Or should the new pass traverse the kernel call-graphs like SAC does?
